[UnifedPDF] Hook up accessibility for Unified PDF

ginalined opened the pull request:
#### 1f1b259fc291228f8aa3a199626845e0529e44bc
<pre>
[UnifedPDF] Hook up accessibility for Unified PDF
Bug 269249 - ([UnifedPDF] Hook up accessibility for Unified PDF)
<a href="https://rdar.apple.com/118903435">rdar://118903435</a> ([UnifedPDF] Hook up accessibility for Unified PDF)

Reviewed by Simon Fraser.

This commit add support so that UnifiedPDF is accessible
This commit create WKAccessibilityPDFDocumentObject as a UnifedPDF accessibility wrapper, which is different from WKPDFPluginAccessibilityObject in that this object does not depends on any layer codes. Previously, the accessibility communication point between webkit and pdfkit is in LayerController. I move the communication point to pdfDocument for better code reuse and less dependency.
Some Todos:
    1. stop accessing the plugin on the accessibility thread
    2. move WKAccessibilityPDFDocumentObject to its own file
    3. resolve coordinate space for better visual experience

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::accessibilityAssociatedPluginParentForElement const): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::pdfDocument const):
(WebKit::PDFPluginBase::didAttachScrollingNode):
(WebKit::PDFPluginBase::activeAnnotation const):
(WebKit::PDFPluginBase::pdfDocumentForPrinting const): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(-[WKAccessibilityPDFDocumentObject initWithPDFDocument:andElement:]):
(-[WKAccessibilityPDFDocumentObject setPDFPlugin:]):
(-[WKAccessibilityPDFDocumentObject setPDFDocument:]):
(-[WKAccessibilityPDFDocumentObject isAccessibilityElement]):
(-[WKAccessibilityPDFDocumentObject accessibilityFocusedUIElement]):
(-[WKAccessibilityPDFDocumentObject document]):
(-[WKAccessibilityPDFDocumentObject accessibilityVisibleChildren]):
(-[WKAccessibilityPDFDocumentObject parent]):
(-[WKAccessibilityPDFDocumentObject setParent:]):
(-[WKAccessibilityPDFDocumentObject accessibilityAttributeValue:]):
(-[WKAccessibilityPDFDocumentObject accessibilityAttributeNames]):
(-[WKAccessibilityPDFDocumentObject accessibilityShouldUseUniqueId]):
(-[WKAccessibilityPDFDocumentObject accessibilityArrayAttributeCount:]):
(-[WKAccessibilityPDFDocumentObject accessibilityChildren]):
(-[WKAccessibilityPDFDocumentObject accessibilityHitTest:]):
(-[WKAccessibilityPDFDocumentObject gotoDestination:]):
(WebKit::PDFPluginBase::PDFPluginBase):
(WebKit::PDFPluginBase::showContextMenuAtPoint):
(WebKit::PDFPluginBase::convertFromPDFViewToRootView const):
(WebKit::PDFPluginBase::convertFromPDFViewToScreenForAccessibility const):
(WebKit::PDFPluginBase::axObjectCache const):
(WebKit::PDFPluginBase::convertFromRootViewToPDFView const):
(WebKit::PDFPluginBase::accessibilityAssociatedPluginParentForElement const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::accessibilityHitTestIntPoint const):
(WebKit::UnifiedPDFPlugin::accessibilityHitTest const):
(WebKit::UnifiedPDFPlugin::accessibilityObject const):
(WebKit::UnifiedPDFPlugin::accessibilityAssociatedPluginParentForElement const): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::pdfDocumentForPrinting const):

Canonical link: <a href="https://commits.webkit.org/274775@main">https://commits.webkit.org/274775@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/003261e269bb54df2abc87886d72d98a85a54e0a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39941 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18952 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42316 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42485 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21859 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16282 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40515 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15971 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34531 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13820 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13850 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43764 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36260 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/35822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39581 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/12137 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16375 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8967 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16019 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,201 @@
+- (PDFDocument*)document;
+- (NSObject *)parent;
+- (void)gotoDestination:(PDFDestination *)destination;
+// - (PDFRect) accessibilityConvertBoundsToRootView:(PDFRect)rect withPageIndex:(NSInteger)pageIndex;
```
We don't check in commented out code.

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,201 @@
+- (void)setPDFDocument:(RetainPtr<PDFDocument>)document;
+- (void)setParent:(NSObject *)parent;
+- (id)accessibilityHitTest:(NSPoint)point;
+- (PDFDocument*)document;
```
Star on the left for C++, right for ObjC, this is ObjC so you're missing a space.

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,201 @@
+@property (assign) WebKit::PDFPluginBase* pdfPlugin;
+
+- (id)initWithPDFDocument:(RetainPtr<PDFDocument>)document andElement:(WebCore::HTMLPlugInElement *)element;
+- (void)setPDFPluginBase:(WebKit::PDFPluginBase *)plugin;
```
And here you have an extra space

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,201 @@
+
+- (NSArray *)accessibilityVisibleChildren
+{
+    NSMutableArray *visiblePageElements = [NSMutableArray new];
```
This is a leak (we don't have ARC on). Adopt it into a RetainPtr and then .autorelease() it at the end.

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,201 @@
+}
+
+ALLOW_DEPRECATED_IMPLEMENTATIONS_BEGIN
+- (id)accessibilityAttributeValue:(NSString *)attribute {
```
Curlies on the next line for ObjC methods (https://webkit.org/code-style-guidelines/)

hortont424 replied:
Repeat this comment many times

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,201 @@
+- (NSArray *)accessibilityAttributeNames {
+ALLOW_DEPRECATED_IMPLEMENTATIONS_END
+    static NeverDestroyed<RetainPtr<NSArray>> attributeNames = @[
+        NSAccessibilityParentAttribute, //
```
what are we `//`ing here?

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,201 @@
+}
+
+- (void)gotoDestination:(PDFDestination *)destination {
+    NSInteger pageIndex = [_pdfDocument indexForPage: [destination page]];
```
no space after :, etc.

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -886,6 +1081,48 @@
+    return handleContextMenuEvent(event);
+}
+
+IntPoint PDFPluginBase::convertFromPDFViewToRootView(const IntPoint& point) const
```
I feel like we decided a while back that "PDFView" coordinate space didn't make sense in the generic case (that it's a PDFPlugin-specific thing). Might this match one of our existing coordinate space conversion methods? If not, I think we need a different name.

ginalined replied:
Current coordinate space is a little off, but I kinda want to solve all the coordinate problems all at once. I think I would push a pr later to solve this problem.

smfr replied:
Please add a FIXME for this here.

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,201 @@
+}
+
+- (NSUInteger)accessibilityArrayAttributeCount:(NSString *)attribute {
+    if (_pdfDocument == nil)
```
No need for explicit nil, we just `if (!_pdfDocument)`

hortont424 replied:
(apply this comment all over)

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,201 @@
 #import "PDFKitSoftLink.h"
 
+
+@interface WKAccessibilityPDFDocumentObject: NSObject {
```
I wonder if we can stick this in its own file.
Also space before the :

ginalined replied:
Will do

ginalined replied:
Take too long to solve all the symbol problems for objective c. Might be a nice to have in the future.

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,201 @@
+#import <WebCore/WebAccessibilityObjectWrapperMac.h>
 #import <wtf/cocoa/VectorCocoa.h>
 #import <wtf/text/TextStream.h>
-
```
This newline was here on purpose (softlink headers have to be in their own paragraph at the end)

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
```diff
@@ -153,7 +154,7 @@ class PDFPluginBase : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<PDF
     virtual id accessibilityHitTest(const WebCore::IntPoint&) const = 0;
     virtual id accessibilityObject() const = 0;
-    virtual id accessibilityAssociatedPluginParentForElement(WebCore::Element*) const = 0;
+    id accessibilityAssociatedPluginParentForElement(WebCore::Element*) const;
```
Not new, but generally webkit coding guidelines say that you should return smart pointers from functions that return ObjC objects. So this would return `RetainPtr<id>`.

ginalined replied:
That would be changing in a lot of places. Do you want me to change it all?

smfr replied:
You can do it in a follow-up.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
```diff
@@ -170,13 +171,21 @@ class PDFPluginBase : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<PDF
+    WebCore::FloatRect convertFromPDFViewToScreen(const WebCore::FloatRect&) const;
+    WebCore::IntPoint convertFromPDFViewToRootView(const WebCore::IntPoint&) const;
+    bool showContextMenuAtPoint(const WebCore::IntPoint&);
+    WebCore::IntPoint convertFromRootViewToPDFView(const WebCore::IntPoint&) const;
```
Switch this and the previous line, add a blank line above `showContextMenuAtPoint()` (for clearer grouping of related functions).

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
```diff
@@ -170,13 +171,21 @@ class PDFPluginBase : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<PDF
+    bool showContextMenuAtPoint(const WebCore::IntPoint&);
+    WebCore::IntPoint convertFromRootViewToPDFView(const WebCore::IntPoint&) const;
+    WebCore::AXObjectCache* axObjectCache() const;
+    void setScrollOffset(const WebCore::ScrollOffset&) final;
```
Group this with the other `ScrollableArea` overrides (i.e. leave it where it was).

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
```diff
@@ -170,13 +171,21 @@ class PDFPluginBase : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<PDF
+    WebCore::AXObjectCache* axObjectCache() const;
+    void setScrollOffset(const WebCore::ScrollOffset&) final;
+#if PLATFORM(MAC)
+    RefPtr<PDFPluginAnnotation> m_activeAnnotation;
```
Don't mix member variables with functions. Move this down with the other member variables (i.e. leave it where it was).

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,208 @@
+}
+
+@property (assign) WeakPtr<WebCore::HTMLPlugInElement, WebCore::WeakPtrImplWithEventTargetData> pluginElement;
+@property (assign) WebKit::PDFPluginBase* pdfPlugin;
```
Can this be a `WeakPtr` too?

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,208 @@
+        if (!focusedElement)
+            [visiblePageElements addObject:page];
+    }
+    return visiblePageElements.get();
```
You're returning a pointer to a deleted object here (since the RetainPtr destructor releases the NSArray).

You need to `-autorelease`

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,208 @@
+
+- (NSObject *)parent
+{
+    RetainPtr<WKAccessibilityPDFDocumentObject> protectedSelf = retainPtr(self);
```
This can just be `RetainPtr protectedSelf = self` I think.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,208 @@
+{
+    RetainPtr<WKAccessibilityPDFDocumentObject> protectedSelf = retainPtr(self);
+    if (!protectedSelf->_parent) {
+        callOnMainRunLoopAndWait([&protectedSelf] {
```
You need to copy the RetainPtr into the captured scope here, so no `&`

Typically we write this as `callOnMainRunLoopAndWait([protectedSelf = RetainPtr { self }] ...`

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,208 @@
+    return visiblePageElements.get();
+}
+
+- (NSObject *)parent
```
If this is always called off the main thread, I'd like to see an `ASSERT(...)` that checks which runloop it's called on, or at least a comment. It needs to be very clear which code is called off the main thread.

ginalined replied:
I think this function is not call from the main thread and that is why I have the `callOnMainRunLoopAndWait` here. I believe I will get bad access if I do not set that up.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,208 @@
+- (NSUInteger)accessibilityArrayAttributeCount:(NSString *)attribute
+{
+    if (!_pdfDocument)
+        _pdfDocument =_pdfPlugin->pdfDocumentForPrinting();
```
Why the "for printing"?

ginalined replied:
I was trying to get the `m_pdfDocument` and I saw a function that returns it 
```
RetainPtr<PDFDocument> pdfDocumentForPrinting() const { return m_pdfDocument; }
```

smfr replied:
We should rename it.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,208 @@
+- (NSArray*)accessibilityChildren
+{
+    if (!_pdfDocument)
+        _pdfDocument =_pdfPlugin->pdfDocumentForPrinting();
```
Ditto.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,208 @@
+    return self;
+}
+
+- (void)gotoDestination:(PDFDestination *)destination
```
Not sure if you can change the name, but it's really unclear what "goto" means here. If it triggers a scroll, we can use the term "reveal" or "scrollTo".

ginalined replied:
this is called by `- (void) scrollToVisible` function in `PDFAccessibilityNodePage.m` and PDFLayerController use the same function name. I agree it is unclear and I will add a comment!

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -104,6 +302,10 @@
     , m_incrementalPDFLoadingEnabled(element.document().settings().incrementalPDFLoadingEnabled())
 #endif
 {
+    m_accessibilityObject = [[WKAccessibilityPDFDocumentObject alloc] initWithPDFDocument:m_pdfDocument andElement:&element];
```
This is a leak; you need to use `adoptNS()`.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -910,6 +1112,48 @@
     return isFullFramePlugin();
 }
 
+// Accessibility Related
```
Can we rename the function rather than require the comment?

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -910,6 +1112,48 @@
+    return valueOrDefault(m_rootViewToPluginTransform.inverse()).mapPoint(pointInPluginCoordinates);
+}
+
+FloatRect PDFPluginBase::convertFromPDFViewToScreen(const FloatRect& rect) const
```
It's really weird for this utility function to call `Accessibility::retrieveValueFromMainThread`. In general, we should not call any functions on PDFPluginBase off the main thread; we don't consider this class to be thread safe at all.

ginalined replied:
I copy this function from PDFPlugin.mm `FloatRect PDFPlugin::convertFromPDFViewToScreen(const FloatRect& rect) const`. Maybe I should not do that?

smfr replied:
It would be OK to rename it `convertFromPDFViewToScreenForAccessibility` for now and leave it, but I think we need to fix this soon.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -910,6 +1112,48 @@
+
+WebCore::AXObjectCache* PDFPluginBase::axObjectCache() const
+{
+    if (!m_frame || !m_frame->coreLocalFrame() || !m_frame->coreLocalFrame()->document())
```
We could use an ASSERT(isMainRunLoop()) here

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -910,6 +1112,48 @@
+    return m_frame->coreLocalFrame()->document()->axObjectCache();
+}
+
+IntPoint PDFPluginBase::convertFromRootViewToPDFView(const IntPoint& point) const
```
`ASSERT(isMainRunLoop())` here too

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -934,6 +1178,17 @@
 
+id PDFPluginBase::accessibilityAssociatedPluginParentForElement(Element* element) const
+{
+    if (!m_activeAnnotation)
```
`ASSERT(isMainRunLoop())`

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1881,19 +1881,20 @@ static IntRect computeMarqueeSelectionRect(const WebCore::IntPoint& point1, cons
-id UnifiedPDFPlugin::accessibilityHitTest(const IntPoint&) const
+id UnifiedPDFPlugin::accessibilityHitTestIntPoint(const WebCore::IntPoint& point) const
 {
-    return nil;
```
`ASSERT(isMainRunLoop())`

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1881,19 +1881,20 @@ static IntRect computeMarqueeSelectionRect(const WebCore::IntPoint& point1, cons
-id UnifiedPDFPlugin::accessibilityObject() const
+id UnifiedPDFPlugin::accessibilityHitTest(const WebCore::IntPoint& point) const
 {
-    return nil;
```
`ASSERT(isMainRunLoop())`

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1881,19 +1881,20 @@ static IntRect computeMarqueeSelectionRect(const WebCore::IntPoint& point1, cons
-id UnifiedPDFPlugin::accessibilityAssociatedPluginParentForElement(Element*) const
+id UnifiedPDFPlugin::accessibilityObject() const
 {
-    return nil;
```
`ASSERT(isMainRunLoop())`

smfr reviewed (changes requested):
The commit message could use some more words; say what's different from PDFPlugin, and describe the threading concerns.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
```diff
@@ -325,6 +334,7 @@ class PDFPluginBase : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<PDF
     uint64_t m_streamedBytes { 0 };
 
     RetainPtr<PDFDocument> m_pdfDocument;
+    RetainPtr<WKAccessibilityPDFDocumentObject> m_accessibilityObject;
```
Maybe call this `m_accessibilityDocumentObject` since it represents the document.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,226 @@
+- (id)initWithPDFDocument:(RetainPtr<PDFDocument>)document andElement:(WebCore::HTMLPlugInElement*)element;
+- (void)setParent:(NSObject *)parent;
+- (void)setPDFDocument:(RetainPtr<PDFDocument>)document;
+- (void)setPDFPluginBase:(WebKit::PDFPluginBase*)plugin;
```
Fine to call this `setPDFPlugin:`

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,226 @@
+
+- (id)accessibilityFocusedUIElement
+{
+    RefPtr plugin = _pdfPlugin.get();
```
Are we on the main thread here? Can we assert that we are?

ginalined replied:
No I think all accessibility functions will not be called from the main thread

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,226 @@
+    if (plugin) {
+        if (RefPtr activeAnnotation = plugin->activeAnnotation()) {
+            if (WebCore::AXObjectCache* existingCache = plugin->axObjectCache()) {
+                if (RefPtr object = existingCache->getOrCreate(activeAnnotation->element()))
```
Put curly braces around the body of the condition, since the macros make it span multiple lines.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,226 @@
+
+- (NSUInteger)accessibilityArrayAttributeCount:(NSString *)attribute
+{
+    if (!_pdfDocument) {
```
This is a bit weird. Why didn't `setPDFDocument:` get called by the plugin?

ginalined replied:
Should be called. But sometimes the pdfdocument is not ready yet when I call it in ‘pluginbase::init’

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,226 @@
+
+- (NSArray*)accessibilityChildren
+{
+    if (!_pdfDocument) {
```
Ditto.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -59,11 +60,226 @@
+- (void)gotoDestination:(PDFDestination *)destination
+{
+    NSInteger pageIndex = [_pdfDocument indexForPage:[destination page]];
+    PDFRect rect = [[self accessibilityChildren][0] accessibilityFrame];
```
I think we should compute the destination scroll position in the main thread code; this computation is almost certainly wrong with rotated pages, zooming etc. We already have `UnifiedPDFPlugin::scrollToPage()` that you can call.

ginalined replied:
added a Fixme and will fix that once I have the separate file for accessibility object

Commit:
[UnifedPDF] Hook up accessibility for Unified PDF
Bug 269249 - ([UnifedPDF] Hook up accessibility for Unified PDF)
rdar://118903435 ([UnifedPDF] Hook up accessibility for Unified PDF)

Reviewed by Simon Fraser.

This commit add support so that UnifiedPDF is accessible
This commit create WKAccessibilityPDFDocumentObject as a UnifedPDF accessibility wrapper, which is different from WKPDFPluginAccessibilityObject in that this object does not depends on any layer codes. Previously, the accessibility communication point between webkit and pdfkit is in LayerController. I move the communication point to pdfDocument for better code reuse and less dependency.
Some Todos:
    1. stop accessing the plugin on the accessibility thread
    2. move WKAccessibilityPDFDocumentObject to its own file
    3. resolve coordinate space for better visual experience

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::accessibilityAssociatedPluginParentForElement const): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::pdfDocument const):
(WebKit::PDFPluginBase::didAttachScrollingNode):
(WebKit::PDFPluginBase::activeAnnotation const):
(WebKit::PDFPluginBase::pdfDocumentForPrinting const): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(-[WKAccessibilityPDFDocumentObject initWithPDFDocument:andElement:]):
(-[WKAccessibilityPDFDocumentObject setPDFPlugin:]):
(-[WKAccessibilityPDFDocumentObject setPDFDocument:]):
(-[WKAccessibilityPDFDocumentObject isAccessibilityElement]):
(-[WKAccessibilityPDFDocumentObject accessibilityFocusedUIElement]):
(-[WKAccessibilityPDFDocumentObject document]):
(-[WKAccessibilityPDFDocumentObject accessibilityVisibleChildren]):
(-[WKAccessibilityPDFDocumentObject parent]):
(-[WKAccessibilityPDFDocumentObject setParent:]):
(-[WKAccessibilityPDFDocumentObject accessibilityAttributeValue:]):
(-[WKAccessibilityPDFDocumentObject accessibilityAttributeNames]):
(-[WKAccessibilityPDFDocumentObject accessibilityShouldUseUniqueId]):
(-[WKAccessibilityPDFDocumentObject accessibilityArrayAttributeCount:]):
(-[WKAccessibilityPDFDocumentObject accessibilityChildren]):
(-[WKAccessibilityPDFDocumentObject accessibilityHitTest:]):
(-[WKAccessibilityPDFDocumentObject gotoDestination:]):
(WebKit::PDFPluginBase::PDFPluginBase):
(WebKit::PDFPluginBase::showContextMenuAtPoint):
(WebKit::PDFPluginBase::convertFromPDFViewToRootView const):
(WebKit::PDFPluginBase::convertFromPDFViewToScreenForAccessibility const):
(WebKit::PDFPluginBase::axObjectCache const):
(WebKit::PDFPluginBase::convertFromRootViewToPDFView const):
(WebKit::PDFPluginBase::accessibilityAssociatedPluginParentForElement const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::accessibilityHitTestIntPoint const):
(WebKit::UnifiedPDFPlugin::accessibilityHitTest const):
(WebKit::UnifiedPDFPlugin::accessibilityObject const):
(WebKit::UnifiedPDFPlugin::accessibilityAssociatedPluginParentForElement const): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::pdfDocumentForPrinting const):

Canonical link: https://commits.webkit.org/274775@main
